tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs

```diff
@@ -0,0 +1,351 @@
+    .maxstack  8
+    IL_0000:  ldc.i4.s   100
+    IL_0002:  newobj     instance void valuetype [runtime]System.Nullable`1<int32>::.ctor(!0)
+    IL_0007:  call       void Test/Thing::Do(valuetype [runtime]System.Nullable`1<int32>)
+    IL_000c:  ret
+  } 
+            """
+            ]))
+
+    [<Test>]
+    let ``Passing an incompatible argument for System.Nullable<'T> method call parameter produces accurate error``() =
+        CompilerAssert.TypeCheckSingleError
+            """
+module Test
+
+type Thing() =
+    static member Do(i: System.Nullable<float>) = ()
+    
+let test() = Thing.Do(true)
+            """
+            FSharpDiagnosticSeverity.Error
+            193
+            (7, 22, 7, 28)
+            """Type constraint mismatch. The type 
+    'bool'    
+is not compatible with type
+    'System.Nullable<float>'    
+"""       
+
+    [<Test>]
+    let ``Assigning a 'T value to a System.Nullable<'T> binding succeeds``() =
+        CompilerAssert.TypeCheckSingleError
+            """
+module Test
+    
+let test(): System.Nullable<int> = 1
+"""
+            FSharpDiagnosticSeverity.Warning
+            3391
+            (4, 36, 4, 37)
+            """This expression uses the implicit conversion 'System.Nullable.op_Implicit(value: int) : System.Nullable<int>' to convert type 'int' to type 'System.Nullable<int>'. See https://aka.ms/fsharp-implicit-convs. This warning may be disabled using '#nowarn "3391"."""
+    
+    [<Test>]
+    let ``Assigning an int32 to a System.Nullable<float> binding fails``() =
+        CompilerAssert.TypeCheckSingleError
+            """
+module Test
+    
+let test(): System.Nullable<float> = 1
+"""
+         FSharpDiagnosticSeverity.Error
+         1
+         (4, 38, 4, 39)
+         """This expression was expected to have type
+    'System.Nullable<float>'    
+but here has type
+    'int'    """
+
+    [<Test>]
+    let ``Overloading on System.Nullable and Result both work without error``() =
+        CompilerAssert.Pass
+            """
+module Test
+    
+type M() =
+    static member A(n: System.Nullable<'T>) = ()
+    static member A(r: Result<'T, 'TError>) = ()
+
+let test() =
+    M.A(System.Nullable 3)
+    M.A(Result<int, string>.Ok 3)
+"""     
+    
+    [<Test>]
+    let ``Overloading on System.Nullable<int>, System.Nullable<'T> and int all work without error``() =
+        CompilerAssert.Pass
+            """
+module Test
+    
+type M() =
+    static member A(n: System.Nullable<'T>) = ()
+    static member A(n: System.Nullable<int>) = ()
+    static member A(n: int) = ()
+
+let test() =
+    M.A(System.Nullable 3)
+    M.A(System.Nullable 3.)
+    M.A(3)
+"""
+
+    [<Test>]
+    let ``Picking overload for typar does not favor any form of System.Nullable nor produce ambiguity warnings``() =
+        CompilerAssert.TypeCheckSingleError
+            """
+module Test
+    
+type M() =
+    static member A(n: System.Nullable<'T>) = ()
+//    static member A(n: System.Nullable<float>) = ()
```

Comment:
Ah, I did want to add, uncommenting this overload does force an ambiguity warning, here and on main, no defaulting to int. Just to confirm, that is what we want? (I also wouldn't quite know how to stop it)

Reply:
I'm a bit surprised by this
* With the extra overload, we get `A unique overload for method 'A' could not be determined`
* Without the extra overload we get a resolution to the `int` overload and `This construct causes code to be less generic than ....`

The same applies if the overload is `Nullable<char>`, e.g. the simpler:
```
type M() =
//    static member A(n: System.Nullable<char>) = printfn "1"
    static member A(n: System.Nullable<int>) = printfn "3"
    static member A(n: int) = printfn "4"
let test(x: 'T) =
    M.A(x)

test 1
```

I can see that extra warnings are being produced in each case and may affect resolution. However I don't understand why the extra overload causes this to happen - why ambiguity in the case of 3 and not 2? It feels like a flaw.

Reply:
Just to be sure there is no confusion, this is not a change that this PR introduced https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzmgFxAEsMAfAgTwAcYACAWQAoBKOgXgFgAoOvu3AQCGBYmDoBbGBOAwodAIJMAdiDoBlSoOkA6AHIBXDBiHAMMADxgAFkKgA+NuzrUoxZQWzK6AIgCMPjz8AsKi4lIycooqapraEvpGJmaW7gSOHC5uHl6+AMyBvPyCImKS0rLySqp0aU5Zabk+KIXmBHQEMIJMCGoA5AAqTkH8DDpKCCxAA

The reason this happens is that when we're in the later stage of overload resolution and we known nothing about the caller arg we fall through to https://github.com/dotnet/fsharp/pull/13673/files#diff-33f19e1cff6e1df09e36df423722449340f685c5c08f041543334c173cc259f8R373
At that point we don't mark these nullable candidates as using TDC and fail to trigger the resolution rule that would end up preferring int over any of the Nullable candidates.

Instead we rely entirely on the ad-hoc compareArg rule for Nullable https://github.com/dotnet/fsharp/blob/main/src/Compiler/Checking/ConstraintSolver.fs#L3208. This one works for the int and Nullable<int> (and even Nullable<'T>) selection case as we can dismiss all the other candidates. This fails to apply when underlying types like char and other non equivalent types are mixed in (with `enforceNullableOptionalsKnownTypes` forcing these to be in the list), leaving no best method.

After fixing that by having these candidates hit the TDC rule instead we get the same method printed twice as a candidate for the test `Ambiguous overload for typar does not pick System.Nullable<'T>` because we now have two warnings for this candidate. 

One being the diag stating it uses a builtin conversion and another stating the typar was not a match.

I can filter out DiagnosticWithText if there is more than one error here https://github.com/dotnet/fsharp/blob/main/src/Compiler/Checking/ConstraintSolver.fs#L3351 or maybe in the driver close to the output is better? (I don't quite understand why candidate method slots need to be duplicated per error).

I could also create a TDC case that produces no warning that will be used in this `enforceNullableOptionalsKnownTypes` branch.

Please advice :)

Reply:
For now I've done the latter https://github.com/dotnet/fsharp/pull/13673/commits/a84386ac0419dcb01a31851d13aa73b29254d122

I still think it's incomplete as I would want to get TDC warnings for the compiler picking a Nullable<'T> overload for 'T as well. But this needs these warning adjustments, I'm not so sure which way to go there.

Reply:
OK, I understand now.  I think don't touch this in this PR - so undo https://github.com/dotnet/fsharp/commit/a84386ac0419dcb01a31851d13aa73b29254d122 - let's get it green and have the primary fixes in the pipe

Reply:
Alright, removed and opened #13731